scipy/sparse/linalg/eigen/lobpcg/lobpcg.py

```diff
@@ -469,80 +491,107 @@ def lobpcg(A, X,
+            myeps = 1e-8
 
-        if verbosityLevel > 0:
-            _report_nonhermitian(gramA, 3, -1, 'gramA')
-            _report_nonhermitian(gramB, 3, -1, 'gramB')
+        if residualNorms.max() > myeps:
+            explicitGramFlag = False
+        else:
+            # suggested by Garrett Moran
+            explicitGramFlag = True
 
-        if verbosityLevel > 10:
-            _save(gramA, 'gramA')
-            _save(gramB, 'gramB')
+        # Shared memory assingments to simplify the code
+        if B is None:
+            blockVectorBX = blockVectorX
+            activeBlockVectorBR = activeBlockVectorR
+            if not restart:
+                activeBlockVectorBP = activeBlockVectorP
+
+        # Common submatrices:
+        gramXAR = np.dot(blockVectorX.T.conj(), activeBlockVectorAR)
+        gramRAR = np.dot(activeBlockVectorR.T.conj(), activeBlockVectorAR)
+
+        if explicitGramFlag:
+            gramXAX = np.dot(blockVectorX.T.conj(), blockVectorAX)
+            gramXBX = np.dot(blockVectorX.T.conj(), blockVectorBX)
+            gramRBR = np.dot(activeBlockVectorR.T.conj(), activeBlockVectorBR)
+            gramXBR = np.dot(blockVectorX.T.conj(), blockVectorBR)
+        else:
+            gramXAX = np.diag(_lambda)
+            gramXBX = ident0
+            gramRBR = ident
+            gramXBR = np.zeros((sizeX, currentBlockSize), dtype=A.dtype)
+
+        if not restart:
+            gramXAP = np.dot(blockVectorX.T.conj(), activeBlockVectorAP)
+            gramRAP = np.dot(activeBlockVectorR.T.conj(), activeBlockVectorAP)
+            gramPAP = np.dot(activeBlockVectorP.T.conj(), activeBlockVectorAP)
+            gramXBP = np.dot(blockVectorX.T.conj(), activeBlockVectorBP)
+            gramRBP = np.dot(activeBlockVectorR.T.conj(), activeBlockVectorBP)
+            if explicitGramFlag:
+                gramPBP = np.dot(activeBlockVectorP.T.conj(),
+                                 activeBlockVectorBP)
+            else:
+                gramPBP = ident
+
+            gramA = bmat([[gramXAX, gramXAR, gramXAP],
+                          [gramXAR.T.conj(), gramRAR, gramRAP],
+                          [gramXAP.T.conj(), gramRAP.T.conj(), gramPAP]])
+            gramB = bmat([[gramXBX, gramXBR, gramXBP],
+                          [gramXBR.T.conj(), gramRBR, gramRBP],
+                          [gramXBP.T.conj(), gramRBP.T.conj(), gramPBP]])
+            if verbosityLevel > 0:
+                _report_nonhermitian(gramA, 3, -1, 'gramA')
+                _report_nonhermitian(gramB, 3, -1, 'gramB')
+            if verbosityLevel > 10:
+                _save(gramA, 'gramA')
+                _save(gramB, 'gramB')
+
+            try:
+                _lambda, eigBlockVector = eigh(gramA, gramB,
+                                               check_finite=False)
+            except LinAlgError:
+                # try again after dropping the direction vectors P from RR
+                restart = True
+
+        if restart:
```

Comment:
Isn't this else of previous `if not restart`

Reply:
> 
> 
> Isn't this else of previous `if not restart`

No, because `restart = True` is redefined a line just above, one can run right after the other